### PR TITLE
Paul/multi partner

### PR DESCRIPTION
**Pull request by paullinator**
### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [ ] Yes
- [ ] No

### Dependencies

 none

### Description

 none

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205469057051686

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import { asArray, asNumber, asObject, asOptional, asString } from 'cleaners'
+
+export const asConfig = asObject({
+  couchDbFullpath: asOptional(
+    asString,
+    'http://username:password@localhost:5984'
+  ),
+  httpPort: asOptional(asNumber, 8008),
+  soloAppIds: asOptional(asArray(asString), undefined),
+  soloPartnerIds: asOptional(asArray(asString), undefined),
```
Comment (samholmes) [inline src/config.ts:11]: Interestingly, cleaner config wont include undefined fields in the output. Perhaps make the default `() => []`?

File: src/config.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  timeoutOverrideMins: asOptional(asNumber, 1200),
+  apiHost: asOptional(asString, undefined),
```
Comment (samholmes) [inline src/config.ts:13]: Suggestion `''` as the default?

File: src/indexApi.ts [hunk 3]
```diff
@@ -81,6 +81,7 @@ async function main(): Promise<void> {
     },
     limit: 1000000
   }
+  console.log(config)
```
Comment (samholmes) [inline src/indexApi.ts:84]: Cleanup?

**Review by samholmes — CHANGES_REQUESTED** [review]

File: src/queryEngine.ts [hunk 4]
```diff
@@ -40,77 +40,47 @@ const asApps = asArray(asApp)
...
+}
+
+const indexes: Index[] = []
+
+INDEXES.forEach(index => {
+  const indexLower = index.map(i => i.toLowerCase())
+  const out = {
+    index: { fields: index },
+    ddoc: indexLower.join('-'),
+    name: indexLower.join('-'),
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/queryEngine.ts]: yep. was redudant

File: src/types.ts [hunk 5]
```diff
@@ -84,7 +84,7 @@ export const asApp = asObject({
   _rev: asString,
   appId: asString,
   appName: asString,
-  pluginIds: asMap(asApiKeys)
+  partnerIds: asMap(asApiKeys)
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/types.ts]: manually in futon

File: src/queryEngine.ts [hunk 6]
```diff
@@ -165,11 +165,13 @@ export async function queryEngine(): Promise<void> {
       // loop over every pluginId that app uses
       remainingPartners = Object.keys(app.partnerIds)
       for (const partnerId in app.partnerIds) {
+        const pluginId = app.partnerIds[partnerId].pluginId ?? partnerId
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/queryEngine.ts:101]: The migration while tedious was simple in that I just had to move an object into a containing block. Adding the pluginId for all of them was looking too tedious plus it's a very special case.

File: src/cacheEngine.ts [hunk 7]
```diff
@@ -14,73 +15,13 @@ const UPDATE_FREQUENCY_MS = 1000 * 60 * 30
...
-          datelog(`${dbName.name} already has '${dbIndex.name}' index.`)
-        } catch {
-          await currentDb.createIndex(dbIndex)
-          datelog(`Created '${dbIndex.name}' index for ${dbName.name}.`)
-        }
-      }
-    }
-  }
+
+  await initDbs()
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/cacheEngine.ts:24]: Why not, only happens at initialization and they don't get created if they exist.

File: src/partners/banxa.ts [hunk 8]
```diff
@@ -20,6 +20,17 @@ import {
...
+export const asBanxaParams = asObject({
+  settings: asObject({
+    latestIsoDate: asOptional(asString, '2018-01-01T00:00:00.000Z')
+  }),
+  apiKeys: asObject({
+    apiKey: asString,
+    secret: asString,
+    partnerUrl: asString
+  })
+})
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/partners/banxa.ts:32]: Possibly but many don't use an Isodate, yet... I'll do that with the next plugin to change.

File: src/partners/banxa.ts [hunk 9]
```diff
@@ -161,31 +174,31 @@ export const banxa: PartnerPlugin = {
...
 
 async function fetchBanxaAPI(
+  partnerUrl: string,
   startDate: string,
   endDate: string,
   pageLimit: number,
   page: number,
-  apiKey: string
+  apiKey: string,
+  secret: string
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/partners/banxa.ts:183]: That's fair. will do

File: src/config.ts (hunk 1 above)
Comment (paullinator) [reply src/config.ts:11]: actually it would need to be `null`

File: src/config.ts (hunk 2 above)
Comment (paullinator) [reply src/config.ts]: null?

File: src/demo/demo.tsx [hunk 10]
```diff
@@ -63,8 +63,8 @@ class App extends Component<
     }
   }
 
-  handleApiKeyChange = (apiKey: any): void => {
-    this.setState({ apiKey: apiKey.target.value })
+  onHandleApiKeyChange = (apiKey: string): void => {
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/demo/demo.tsx:68]: This is actually an eslint rule in the project.
Screenshot 2024-02-09 at 22 42 11

File: src/config.ts (hunk 2 above)
Comment (paullinator) [reply src/config.ts]: changed to `http://localhost:8008`

**Pull request merged**